src/sdsc/xslt.py

```diff
@@ -0,0 +1,396 @@
+        filename = "<file>%s</file>" % str(basefile)
+
+    withinid = ""
+    if contextid:
+        withinid = "<withinid>%s</withinid>" % str(contextid)
+
+    return etree.XML("""<result type="error">
+            <location>%s%s<line>%s</line></location>
+            <message><quote>%s</quote> is duplicated:
+                <quote>%s</quote>
+            </message>
+            <suggestion>Remove one instance of <quote>%s</quote>.</suggestion>
+        </result>""" % (filename, withinid, str(linenumber(context)), duplicate, quote, duplicate))
+
+
+# --------------------------------------------------------------------
+# XSLT Extension Functions
+# --------------------------------------------------------------------
+
+
+def linenumber(context):
+    """Returns the line number of the current context
+
+    :param str context: Context node (ignored)
+    """
+    return context.context_node.sourceline
+
+
+def splitpath(context, path, wantedsegment='filename'):
+    """Split a path into segments and return the wanted segment.
+
+    :param str context: Context node (ignored)
+    :param str path: Input path or file name
+    :param str wantedsegment: Segment that you want as output
+        ('path' => path without file name and without trailing slash,
+        'filename' => filename without extension,
+        'extension' => file extension with leading dot and in lowercase)
+    """
+
+    if not path:
+        return []
+
+    # I get this as a list with one lxml.etree._ElementUnicodeResult, not
+    # as a list with a string.
+    path = str(path[0])
+
+    splitpath = os.path.split(path)
+    splitfile = os.path.splitext(splitpath[1])
+
+    if wantedsegment == 'path':
+        return splitpath[0]
+    elif wantedsegment == 'filename':
+        return splitfile[0]
+    elif wantedsegment == 'extension':
+        return splitfile[1].lower().lstrip('.')
+    else:
+        return path
+
+
+def counttokens(context, text):
+    """TODO@sknorr:
+
+    :param str context: Context node (ignored)
+    """
+
+    count = 0
+    if text:
+        count = len(tokenizer(text[0]))
+    return count
+
+
+def tokenizer(text):
+    """TODO@sknorr: Splits text into words
+
+    :param str text:
+    """
+    return text.split()
+
+
+def sentencesegmenter(text):
+    """TODO@sknorr:
+
+    :param str text:
+    """
+    sentences = sentenceends.split(text)
+    # The last sentence's final punctuation has not yet been cut off, do that
+    # now.
+    sentences[-1] = lastsentenceends.sub('', sentences[-1])
+
+    # We also need to cut off parentheses etc. from the first word of the first
+    # sentence, as that has not been done so far either.
+    sentences[0] = removepunctuation(sentences[0], start=True)
+    return sentences
+
+
+def sentencelengthcheck(context, content, contentpretty, contextid, basefile,
+                        lengthwarning, lengtherror):
+    # Try to use sensible defaults. The following seems like better advice than
```

Comment:
"Takes a paragraph, splits up sentences and checks whether the number of words in the sentences is longer than a defined maximum."